APM Metrics, UI Baseline

jasonrhodes opened the pull request:
## Summary

Step 1: Refactor APM transaction/error overview pages into a single service details page

To make room for the new Metrics tab, this PR refactors the existing Transaction and Error Group overview pages into a new ServiceDetails page, with shared UI moved up and out of those components, along with a good amount of TS conversion.

Step 2: Add metrics tab to the service details page, with the following charts:
* CPU line chart
* Memory area chart
* Error occurrences bar chart
* Standard transaction charts for duration and TPM (service-wide, all transaction types)
  * Standard transaction charts for each transaction type, also

TODO
* [x] activate the integrations button with both the ML and the Watcher integrations in a single context menu
* [x] check tests
* [x] fix possible bug with transaction type filtering via the kuery bar (why does 2nd type never show up at all?)
* [x] add transaction type filter dropdown on the transactions tab
* [x] use transaction type filter dropdown in ML flyout (default to context or first type)
* [x] ML annotations / label and tooltip
* [x] convert metric names to constants
* [x] handle "no hits" scenario on new metrics charts
* [x] confirm charts are resilient if a metric series doesn't exist (query-side)

jasonrhodes committed:
Service detail page refactor with TS conversions

jasonrhodes committed:
Adds error group list back to the service details page tabs

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2213/)

jasonrhodes committed:
Fixes after rebasing upstream master

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2215/)

sqren commented on x-pack/plugins/apm/common/constants.ts:
```diff
@@ -41,3 +43,8 @@ export const ERROR_LOG_STACKTRACE = 'error.log.stacktrace';
+export const TRANSACTION_TYPE_MAP: StringMap<string> = {
+  request: 'Request',
+  'page-load': 'Page load'
+};
```
This doesn't really fit in here, does it? All the others are constants for ES properties.

jasonrhodes committed:
Adds integrations menu and flyouts, with no-transaction ML job creation and lookup

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2430/)

ogupte commented on x-pack/plugins/apm/public/components/app/ServiceDetails/ServiceIntegrations/MachineLearningFlyout.tsx:
```diff
@@ -0,0 +1,256 @@
+interface FlyoutState {
+  isLoading: boolean;
+  hasMlJob: boolean;
+  hasIndexPattern: boolean | null;
```
This doesn't actually need to be `null`. It looks like the render method only checks that `hasIndexPattern` is falsy, so it could be initialized as `false` and could avoid the awkward tertiary nature of this union type.

ogupte commented on x-pack/plugins/apm/public/components/app/ServiceDetails/ServiceIntegrations/MachineLearningFlyout.tsx:
```diff
@@ -0,0 +1,256 @@
+          {hasMlJob && (
+            <div>
+              <EuiCallOut
+                title={<span>Job already exists</span>}
```
Why is the `<span>` necessary?

sqren replied:
I think it's required by EUI. Would be nice if simple `string` types were also allowed.

ogupte replied:
In the EUI docs, the propType is listed as `node` and the example code only shows a string being passed in for `title`

sqren replied:
Okay, maybe I'm wrong. I've just previously experienced the need for wrapping in `Fragment` et al.

jasonrhodes replied:
EuiTitle and possibly EuiText have some sort of requirement along these lines, but this one may not. This component was copied/pasted over so I'm not sure why the span was there originally.

ogupte commented on x-pack/plugins/apm/public/components/app/ServiceDetails/ServiceIntegrations/MachineLearningFlyout.tsx:
```diff
@@ -0,0 +1,256 @@
+            {/* <img
+              src="/plugins/apm/images/apm-ml-anomaly-detection-example.png"
+              alt="Anomaly detection on response times in APM"
+            /> */}
```
Should the commented code blocks be removed? if not, maybe add a note/todo.

sqren commented on x-pack/plugins/apm/public/components/app/ServiceDetails/ServiceIntegrations/MachineLearningFlyout.tsx:
```diff
@@ -0,0 +1,256 @@
+  };
+
+  public componentDidMount() {
+    getAPMIndexPattern().then((indexPattern: string | null) => {
```
`getAPMIndexPattern` returns an indexPattern object (not a string). And I think it returns `undefined` (not `null`) when it cannot be found.

Btw. I have [a PR ready](https://github.com/elastic/kibana/pull/26622) that converts rest/`savedObjects.ts` to ts and changes the output a bit (shouldn't affect you here).

sqren commented on x-pack/plugins/apm/public/components/app/ServiceDetails/ServiceIntegrations/MachineLearningFlyout.tsx:
```diff
@@ -0,0 +1,256 @@
+  public state = {
+    isLoading: false,
+    hasIndexPattern: null,
+    hasMlJob: false
```
Now we have `mlAvailable`, `hasMlJob` and `hasDynamicBaseline`. I can't remember if they are all truly different, or if they overlap/are the same.

jasonrhodes replied:
yeah, `mlAvailable` is about the license, `hasMlJob` is about if the current context (service + transaction type) already has a job created. `hasDynamicBaseline` is being replaced with `hasMlJob` :+1:

sqren commented:
I assume you'll delete `TransactionOverview/DynamicBaseline/` and `ErrorGroupOverview/Watcher/`?

jasonrhodes commented on x-pack/plugins/apm/public/components/app/ServiceDetails/index.tsx:
```diff
@@ -0,0 +1,25 @@
+
+const ServiceDetails = connect(mapStateToProps)(ServiceDetailsView);
+
+export { ServiceDetails };
```
Heh, I actually hate it, but Chandler and I discovered through a ton of trial and error that this is the only way for the export to inherit the variable name as the "function.name" value, and then get pulled in as React's "displayName". In other words, doing it like this means it will show up as `Connect<ServiceDetails>` in snapshots but doing it like `export const ServiceDetails =` will mean it shows up as `Connect<Component>` :/

sqren replied:
Oooh, that's interesting!

delvedor commented on x-pack/plugins/apm/server/lib/transactions/get_top_transactions/index.ts:
```diff
@@ -10,29 +10,42 @@ import {
   const { start, end } = setup;
 
-  const bodyQuery = {
+  // TODO: Find an ES type for this instead
```
I'll work on that soon! A first types implementation has landed in `next`, the next step is improving the support for typed responses.

jasonrhodes commented on x-pack/plugins/apm/common/constants.ts:
```diff
@@ -41,3 +43,8 @@ export const ERROR_LOG_STACKTRACE = 'error.log.stacktrace';
+export const TRANSACTION_TYPE_MAP: StringMap<string> = {
+  request: 'Request',
+  'page-load': 'Page load'
+};
```
Hm, I think if we have a `/common/constants` file it should probably be the place for any constant we want to define? If we want it to be an ES property file, we could rename and then create a more general constants file next to it. I'm not sure if i have buy in yet to remove this name forking overall, but if I do, then we can remove this altogether. If we need to keep it, I'll figure out where this can go.

sqren replied:
True, a rename to something specific might be a better idea. 
That being said, I don't think `TRANSACTION_TYPE_MAP` is actually a constant. It's a utility method for converting `transactionType` to a ui label, and is used as a dynamic lookup `TRANSACTION_TYPE_MAP[transactionType]` whereas `ERROR_EXC_HANDLED` is a constant used across the codebase.

sqren replied:
Btw to make matters worse we also have https://github.com/elastic/kibana/blob/948f4bc6bb306fc076faa25652a07044f6950582/x-pack/plugins/apm/public/constants/index.js 😱

sqren replied:
> Maybe having some more localized components/SomeComponent/constants.js could work for these kind of values

Yeah, I think that makes more sense for ui stuff like `TRANSACTION_TYPE_MAP` since it's closely tied to where it's used. We for instance have other methods that maps transaction types to other labels.
How are you handling values other than the ones in the map? Would it be an idea to use a switch statement with a default value?

jasonrhodes replied:
Cool, I removed them for now and will revisit if we need to re-implement this kind of relabeling after talking with the team tomorrow at the weekly.

jasonrhodes commented on x-pack/plugins/apm/public/components/shared/charts/CustomPlot/StaticPlot.js:
```diff
@@ -89,7 +89,14 @@ class StaticPlot extends PureComponent {
+          tickValues={yTickValues}
+          tickFormat={tickFormatY}
+          style={{
+            line: { stroke: 'none', fill: 'none' }
```
Yeah, in Casper's mock ups the left-side Y-axis line is gone now that the charts are inside of panels. I *think* that is what this does, IIRC. Will check again.

makwarth replied:
Let's talk about it during next UI call?

sqren replied:
Sounds good 👍

jasonrhodes commented on x-pack/plugins/apm/public/components/app/ErrorGroupOverview/beta.tsx:
```diff
@@ -0,0 +1,30 @@
+/*
```
Just as I work through these changes, to make it easy to revert some things if I need to. Will switch it out with index once we've settled.

jasonrhodes commented on x-pack/plugins/apm/public/components/app/ErrorGroupOverview/view.js:
```diff
@@ -8,12 +8,13 @@ import React, { Component } from 'react';
 import { get } from 'lodash';
 import { HeaderContainer } from '../../shared/UIComponents';
-import TabNavigation from '../../shared/TabNavigation';
+import { TransactionErrorTabs } from '../../shared/TransactionErrorTabs';
```
This file will go away once the stuff in beta.tsx is finalized. That TransactionErrorTabs file is gone as I opted to render the tabs in the service details page instead of sharing them across pages.

jasonrhodes commented on x-pack/plugins/apm/public/services/rest/ml.ts:
```diff
@@ -0,0 +1,98 @@
+}) {
+  const indexPatternName = chrome.getInjected('apmIndexPatternTitle');
+  const groups = ['apm', serviceName.toLowerCase()];
+  const filter: TermFilterSet = [{ term: { [SERVICE_NAME]: serviceName } }];
```
Oh right, great idea! This was bothering me…

jasonrhodes committed:
WIP adds memory usage chart with fake data

jasonrhodes committed:
Adds error group distribution chart to metrics tab

jasonrhodes committed:
Metrics "memory" chart data flow from server to chart

jasonrhodes committed:
Server side query and transform set up for cpu charts

jasonrhodes committed:
CPU charts created

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3039/)

sqren commented on x-pack/plugins/apm/server/lib/metrics/query_types.ts:
```diff
@@ -0,0 +1,24 @@
+  serviceName: string;
+  setup: Setup;
+  timeseriesAggregates?: StringMap<any>;
+  otherAggregates?: StringMap<any>;
```
I've only heard it referred to as "aggregations". So I think the above two should be `timeseriesAggregations` and `otherAggregates.`
Btw. what do you think of `bucketAggregations` and `totalAggregations` or something like that to distinguish that the first set is run per histogram bucket, the other set is global?

sqren commented on x-pack/plugins/apm/server/lib/metrics/get_cpu_chart_data/transformer.ts:
```diff
@@ -0,0 +1,75 @@
+    systemCPUAverage: AggValue['value'];
+    systemCPUMax: AggValue['value'];
+    processCPUAverage: AggValue['value'];
+    processCPUMax: AggValue['value'];
```
`AggValue['value']` makes it a bit harder to immediately see what's going on where `number | null` is more readable.

jasonrhodes committed:
Review feedback and watcher flyout

jasonrhodes commented on x-pack/plugins/apm/public/components/app/ServiceDetails/view.tsx:
```diff
@@ -0,0 +1,84 @@
+
+        <KueryBar />
+
+        <HistoryTabs tabs={tabs} contentWrapper={TabContentWrapper} />
```
This is removed in favor of a hard-coded EuiSpacer in the HistoryTabs component. (I was able to remove spacers from the traces and service overview components that were there for this same reason.)

jasonrhodes commented on x-pack/plugins/apm/public/components/shared/UIComponents.ts:
```diff
@@ -28,19 +27,19 @@ export const HeaderLarge = styled.h1`
   margin: ${px(units.plus)} 0;
   font-size: ${fontSizes.xlarge};
   ${props => props.css};
 `;
```
Almost there!

jasonrhodes committed:
Server tests added, broken tests fixed

jasonrhodes committed:
Review feedback

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3318/)

jasonrhodes committed:
Fixes up type problems

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3323/)

jasonrhodes committed:
Converts metric key paths to constants

jasonrhodes committed:
Fixes error group type issue after rebasing from master

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3419/)

sqren commented on x-pack/plugins/apm/common/__snapshots__/constants.test.ts.snap:
```diff
@@ -194,6 +230,18 @@ exports[`Transaction v2 ERROR_LOG_MESSAGE 1`] = `undefined`;
+exports[`Transaction v2 METRIC_SYSTEM_FREE_MEMORY 1`] = `undefined`;
+
+exports[`Transaction v2 METRIC_SYSTEM_TOTAL_MEMORY 1`] = `undefined`;
+
```
Seems like we need a better solution here. I'll be adding two additional sections for error constants also (v1, and v2), and that'll duplicate the metric constants twice more.
Anything in mind?

jasonrhodes replied:
I almost tried to solve it and then backed away haha, but yeah I'm with you. We probably need to test groups of config against mock documents that have the expected key paths ... we are doing it only against transactions and spans currently. I wasn't sure if we wanted a "Metric" type for the document or just an ad hoc mock metric doc... etc

sqren commented on x-pack/plugins/apm/server/lib/helpers/setup_request.ts:
```diff
@@ -36,7 +40,7 @@ export type ESClient = <T = void, U = void>(
   start: number;
   end: number;
-  esFilterQuery?: any;
+  esFilterQuery?: ESFilter;
```
👍

sorenlouv reviewed (commented):
This looks really good. Good job 👍 
Please make sure to manually test the areas you've touched notably Watcher and ML integration.

jasonrhodes commented on x-pack/plugins/apm/public/components/app/ServiceDetails/ServiceIntegrations/WatcherFlyout.tsx:
```diff
@@ -57,58 +56,87 @@ const SmallInput = styled.div`
   };
 
-  onChangeThreshold = event => {
+  public onChangeThreshold = (event: DOMEvent) => {
```
oh thank god, perfect :pray:

jasonrhodes commented on x-pack/plugins/apm/public/components/app/ServiceDetails/ServiceIntegrations/MachineLearningFlyout.tsx:
```diff
@@ -4,59 +4,103 @@
+  public addErrorToast = () => {
+    const {
+      serviceName,
+      transactionType = 'all transactions',
```
Possibly yeah, I'm going to revert all of this "all transactions" stuff

jasonrhodes committed:
Makes transaction type filterable on transactions list tab

jasonrhodes commented on src/legacy/core_plugins/apm_oss/index.js:
```diff
@@ -34,6 +34,7 @@ export default function apmOss(kibana) {
         onboardingIndices: Joi.string().default('apm-*'),
         spanIndices: Joi.string().default('apm-*'),
         transactionIndices: Joi.string().default('apm-*'),
+        metricsIndices: Joi.string().default('apm-*'),
```
Yeah I think the plural is ok here, good call on adding to documentation -- will figure that out as I wrap this up.

jasonrhodes commented on x-pack/plugins/apm/public/components/app/ServiceDetails/ServiceIntegrations/createErrorGroupWatch.ts:
```diff
@@ -4,36 +4,49 @@
+  schedule: any; // TODO: what is this type?
+  serviceName: string;
+  slackUrl?: string | null;
+  threshold: any; // TODO: what is this type?
```
thx!

jasonrhodes commented on x-pack/plugins/apm/public/utils/formatters.ts:
```diff
@@ -104,17 +104,21 @@ export function asInteger(value: number) {
   return numeral(numerator / denominator).format('0.00%');
 }
+
+export function asGB(bytes: number | null, places: number = 1) {
```
If `bytes` is `null`, the default value does nothing. And the ES value that comes back in this case is `null`, so we have to handle that one way or another 🤔

jasonrhodes commented on x-pack/plugins/apm/server/testHelpers/mocks.ts:
```diff
@@ -0,0 +1,29 @@
+import { Setup } from '../lib/helpers/setup_request';
+
+interface TestSetup extends Setup {
+  client: any; // tried to use jest's Mock type here but it wouldn't import
```
Ahhh wow yes

jasonrhodes committed:
Review feedback

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3762/)

jasonrhodes commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3828/)

jasonrhodes committed:
Chart formatting tweaks and fixes

jasonrhodes committed:
Fixes y-axis ticks for GB chart

jasonrhodes committed:
Restores machine learning annotations with some test fixes

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4019/)

jasonrhodes committed:
Continued styling and other review based changes

jasonrhodes commented on x-pack/plugins/apm/public/components/app/ServiceDetails/CPUUsageChart.tsx:
```diff
@@ -0,0 +1,41 @@
+
+const CPUUsageChart: React.SFC<Props> = ({ data }) => (
+  <SyncChartGroup
+    render={syncProps => (
```
Gonna just leave this as-is for now but we can change the name later if that makes sense. :)

jasonrhodes commented on x-pack/plugins/apm/public/components/app/ServiceDetails/ServiceIntegrations/MachineLearningFlyout.tsx:
```diff
@@ -4,66 +4,106 @@
+      transactionType
     });
+    this.setState({ hasMLJob: count > 0 });
   }
```
I think this is how we determined this before:
https://github.com/elastic/kibana/blob/47c163b2f5642bbe15c2f14218bcd5c4e8b528b4/x-pack/plugins/apm/public/store/reactReduxRequest/transactionOverviewCharts.tsx#L43-L47

which, IIUC, assumed the transaction charts had already been fetched. We now need to know if a service/type combination has an ML job outside of when those charts are fetched (e.g. in the machine learning flyout, we need to know all of the service + type combinations and whether they have ML jobs attached). 

I already converted this to RRR in another place, let me see if I can easily use the same request component here or not.

sqren replied:
> I already converted this to RRR in another place, let me see if I can easily use the same request component here or not.

👍

jasonrhodes replied:
Actually now that I understand this better, *this* is the place where I'd already converted it to RRR so it's already good. The other place I am doing this lookup is in the transaction charts component, and in that case we already have the chart data so I can use the old lookup there and avoid these network calls. Changing that now :)

jasonrhodes commented on x-pack/plugins/apm/public/components/app/ServiceDetails/ServiceIntegrations/MachineLearningFlyout.tsx:
```diff
@@ -76,8 +116,12 @@ export default class DynamicBaselineFlyout extends Component {
+  public addSuccessToast = () => {
+    const {
+      serviceName,
+      transactionType = 'all transactions',
```
Thanks, good catch

jasonrhodes commented on x-pack/plugins/apm/public/components/app/ServiceDetails/ServiceDetailTabs.tsx:
```diff
@@ -0,0 +1,57 @@
+  location: any;
+}
+
+export class Tabs extends React.PureComponent<TabsProps> {
```
Hm, I should've written a comment when I did this. I had a reason but I can't remember it now. I'm going to switch to regular `Component` for now unless I can recall.

jasonrhodes commented on x-pack/plugins/apm/public/components/app/ServiceDetails/MemoryUsageChart.tsx:
```diff
@@ -0,0 +1,41 @@
+  />
+);
+
+export { MemoryUsageChart };
```
no I don't think so -- is there a way to do that and type it as `React.SFC<Props>` ?

jasonrhodes replied:
>is there a way to do that and type it as React.SFC<Props>?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4324/)

jasonrhodes committed:
More review fixes and cleanup

jasonrhodes committed:
Updated chartGroupProps -> hoverXHandlers

sqren commented on x-pack/plugins/apm/public/store/reactReduxRequest/license.tsx:
```diff
@@ -4,26 +4,31 @@
     ml: { isAvailable: false }
   },
-  license: { isActive: false }
+  license: { is_active: false }
```
I think `callApi` still camelCases response objects automatically, so not sure this is right.
If that's the case you can disable camecasing for just the license endpoint. In a follow-up PR I want to disable it by default, and only have i camelcase the places that need it. And they should eventually be migrated to not use the camelcased props.

sqren commented on x-pack/plugins/apm/public/services/rest/ml.ts:
```diff
@@ -0,0 +1,91 @@
+}
+
+// https://www.elastic.co/guide/en/elasticsearch/reference/6.5/ml-get-job.html
+export interface GetMLJobApiResponse {
```
Nit: For the other interfaces we haven't prefixed with a verb (start, get). To make it consistent we could name them like:

```
StartMLJobApiResponse -> StartedMLJobApiResponse
GetMLJobApiResponse -> MLJobApiResponse
```

I think that makes sense if you think about what is returned vs what action caused it.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4338/)

jasonrhodes commented on x-pack/plugins/apm/public/components/app/ErrorGroupOverview/index.tsx:
```diff
@@ -0,0 +1,30 @@
+  location: any;
+}
+
+export const ErrorGroupOverview: React.SFC<ErrorGroupOverviewProps> = ({
```
AIUI yes. I definitely have not caught all of these.

jasonrhodes committed:
Removes "type" column from transaction list table

sqren commented on x-pack/plugins/apm/public/components/app/ServiceDetails/view.tsx:
```diff
@@ -0,0 +1,72 @@
+export class ServiceDetailsView extends React.Component<ServiceDetailsProps> {
+  public render() {
+    const { urlParams, match, location } = this.props;
+    const { serviceName } = match.params;
```
`serviceName` is available on `urlParams`

sqren commented on x-pack/plugins/apm/public/components/app/ServiceDetails/view.tsx:
```diff
@@ -0,0 +1,72 @@
+    const { urlParams, match, location } = this.props;
+    const { serviceName } = match.params;
+
+    const params = { ...urlParams, serviceName };
```
Not necessary since `serviceName` is already part of `urlParams`

sqren commented on x-pack/plugins/apm/public/components/app/ServiceDetails/ServiceIntegrations/view.tsx:
```diff
@@ -0,0 +1,163 @@
+      this.addMlPanelItems();
+    }
+    this.addWatcherPanelItems();
+  };
```
Personally I find that getters are easier to read - I don't have to lookup what side-effect the method I call has:

```ts
public getPanelItems = () => {
    const panelItems = [];
    if (this.props.mlAvailable) {
      panelItems.push(...this.getMlPanelItems());
    }
    panelItems.push(...this.getWatcherPanelItems());
    return panelItems;
  };
```

sqren commented on x-pack/plugins/apm/public/components/app/ServiceDetails/ServiceIntegrations/view.tsx:
```diff
@@ -0,0 +1,163 @@
+            panels={[
+              {
+                id: 0,
+                items: this.panelItems
```
Pretty sure this is not recommended. If `this.panelItems` change outside a render cycle, the component won't re-render, and the user will see stale data.

Instead it should be `this.state.panelItems` - or even simpler: drop the lifecycle methods and calculate it in the render function. Concatenating small arrays like this is fast and won't cause perf issues.

```jsx
items: this.getPanelItems()
```

sqren commented on x-pack/plugins/apm/public/components/app/ServiceDetails/ServiceMetrics.tsx:
```diff
@@ -0,0 +1,112 @@
+  const params = { serviceName, ...urlParams };
+  if (serviceTransactionTypes.length !== 1) {
+    delete params.transactionType;
+  }
```
I think this deserves a comment.

sqren commented on x-pack/plugins/apm/public/components/app/ServiceDetails/ServiceMetrics.tsx:
```diff
@@ -0,0 +1,112 @@
+  serviceName,
+  urlParams,
+  serviceTransactionTypes,
+  chartWrapper: ChartWrapper = React.Fragment
```
Is this still used? If not it should be removed

sqren commented on x-pack/plugins/apm/public/components/app/TransactionOverview/index.tsx:
```diff
@@ -0,0 +1,116 @@
+      return null;
+    }
+
+    const updatedParams = { ...urlParams, transactionType };
```
Why is this necessary?

sqren commented on x-pack/plugins/apm/public/components/app/TransactionOverview/index.tsx:
```diff
@@ -0,0 +1,116 @@
+            <TransactionCharts
+              charts={data}
+              location={location}
+              urlParams={updatedParams}
```
The idea with `urlParams` is that they are passed top-down throughout the app, and stay the same. It can be really confusing if they start differing between components.

sqren commented on x-pack/plugins/apm/public/components/shared/charts/CustomPlot/InteractivePlot.js:
```diff
@@ -34,16 +34,21 @@ class InteractivePlot extends PureComponent {
       });
+
+    // sort tooltip values in DESC order so they match the series
+    points.sort((a, b) => (a.point.y < b.point.y ? 1 : -1));
```
The tooltip values should already follow the order of the series...

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4374/)

jasonrhodes committed:
Re-centers around using urlParams

jasonrhodes committed:
Fixes panel item references in service integrations menu

jasonrhodes commented on x-pack/plugins/apm/public/components/app/ServiceDetails/ServiceIntegrations/MachineLearningFlyout.tsx:
```diff
@@ -0,0 +1,280 @@
+          this.props.transactionType || this.props.serviceTransactionTypes[0]
+      });
+    }
+  }
```
I can only change this value if transaction type changes, otherwise I want to use the value in state that is otherwise controlled by the select box. I'm sure there's a better way to do this but I'm not seeing it yet, punting for now.

sqren replied:
Sounds good. Let's pick the importing battles for now. 👍

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4440/)

jasonrhodes committed:
Serves default transaction type from /transactions route

jasonrhodes committed:
Reverts tooltip series ordering change

jasonrhodes committed:
Adds a clarification note about route config

jasonrhodes committed:
Makes it possible to request transaction chart data without specifying type

jasonrhodes committed:
Fixes empty series problem for memory chart

jasonrhodes committed:
Fixes tests

jasonrhodes commented:
OK @sqren, I've changed/fixed/updated everything we talked about earlier today and added a new endpoint in the back end to return all transaction chart data, regardless of type. (The difference is hard to see in opbeans, because request is by far the dominant type, but it seems to work.)

Relevant commits:
* Re-centers everything around urlParams: https://github.com/elastic/kibana/pull/26608/commits/73441fdc501a771ad6ccb53badbb95e049acb873
* Removes `this.panelItems` anti-pattern: https://github.com/elastic/kibana/pull/26608/commits/c12a6068cab81f07872cd5a4e907799bba5e56a2
* Makes /transactions show the default transaction type without a redirect: https://github.com/elastic/kibana/pull/26608/commits/b4a4788292e384cc85b49c9fd85353b357fc5649
* Reverts the tooltip series ordering change: https://github.com/elastic/kibana/pull/26608/commits/1a78fa26271bb747cca336e51d66455661f32cf6
* New backend route and RRR request type to get transaction chart data without a transaction type: https://github.com/elastic/kibana/pull/26608/commits/93f2694e035ffdcbd8cd77c36e616ca79db5ad2d
* Fixes empty series problem with new charts (also solves NaN errors in console): https://github.com/elastic/kibana/pull/26608/commits/09603a485d00628064ca5fe68b084bf9ba9c357d

I'll spend some time thinking this evening about if/how we can split this PR up, or possibly we can just start with a new PR and go over it again together when I'm around tomorrow?

sqren commented on x-pack/plugins/apm/public/store/reactReduxRequest/transactionOverviewCharts.tsx:
```diff
@@ -69,3 +72,26 @@ export function TransactionOverviewChartsRequest({ urlParams, render }: Props) {
+      render={render}
+    />
+  );
+}
```
Next iteration we can hopefully clean up the `transactionType` implicitness, and add the appropriate redirects as we talked about, and then get rid of this and just use `TransactionOverviewChartsRequest`.

jasonrhodes committed:
Review feedback

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4483/)

sqren commented on x-pack/plugins/apm/public/components/app/ServiceDetails/ServiceIntegrations/MachineLearningFlyout.tsx:
```diff
@@ -0,0 +1,287 @@
+          }
+
+          const hasMLJob = Boolean(
+            data.jobs.find(
```
Nit: `data.jobs.some` would probably be more appropriate.

jasonrhodes replied:
Good call

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4492/)

jasonrhodes committed:
Fixes after rebase

jasonrhodes committed:
Replaces boolean casted array#find with array#some

jasonrhodes commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4555/)

sqren commented on x-pack/plugins/apm/public/components/app/TransactionOverview/List/index.js:
```diff
@@ -70,10 +69,10 @@ export default function TransactionList({
     {
       field: 'transactionsPerMinute',
-      name: tpmLabel(type),
+      name: 'Trans. per minute',
```
@jasonrhodes was this resolved?
